File: webviewerwidgets/wgsosamplesummary/wgsosamplesummary.js
widgetGenerators['sosamplesummary'] = {
	'info': {
		'name': 'Sequence Ontology by Sample',
		'width': 480, 
		'height': 380, 
		'callserver': true,
        'variables': {},
        'init': function (data) {
            this['variables']['data'] = data;
        },
        'shoulddraw': function () {
            var data = this['variables']['data'];
            if (data == null) {
                return false;
            } else {
                if (data['samples'].length > 20) {
                    return false;
                } else {
                    return true;
                }
            }
        },
        'beforeresize': function () {
            var contentDiv = this['variables']['div'];
            $(contentDiv).empty();
            contentDiv.style.width = '0px';
            contentDiv.style.height = '0px';
        },
        'onresize': function () {
            var v = this['variables'];
            this['function'](v['div'], v['data']);
        },
		'function': function (div, data) {
			var colorPalette = {
				'Frameshift insertion':'#ff0000', // red
				'Frameshift deletion':'#dc143c', // crimson
				'Stopgain':'#ff4500', // orange red
				'Stoploss':'#ffa500', // orange
				'Missense':'#ffd700', // gold
				'Inframe insertion':'#d2691e', // chocolate
				'Inframe deletion':'#8b4513', // saddle brown
				'Splice site':'#adff2f', // green yellow
				'2k upstream':'#7fffd4', // aqua marine
				'2k downstream':'#00ced1', // dark turquoise
				'3\' UTR':'#00bfff', // deep sky blue
				'5\' UTR':'#ffff00', // yellow
				'Complex substitution':'#0000ff', // blue
				'Synonymous':'#00ff00', // lime
				'Intron':'#00ffff', // aqua
				'Unknown':'#000080', // navy
				'Intergenic':'#008080', // teal
	        };
			if (div != null) {
				emptyElement(div);
			}
            this['variables']['div'] = div;
            this['variables']['data'] = data;
            div.style.position = 'relative';
            div.style.overflowX = 'scroll';
            var divWidth = $(div).width();
            $(div).width(divWidth);
            var sdiv = getEl('div');
            sdiv.style.position = 'relative';
            sdiv.style.width = '100%';
            sdiv.style.height = '100%';
            addEl(div, sdiv);
			var chartDiv = getEl('canvas');
            var canvasWidth = data['samples'].length * 30 + 100;
            var canvasHeight = div.clientHeight;
            chartDiv.style.position = 'absolute';
            chartDiv.style.top = '0';
            chartDiv.style.left = '0';
            chartDiv.style.pointerEvents = 'none';
			//chartDiv.style.width = canvasWidth + 'px';
			//chartDiv.style.height = '200px';
            //chartDiv.width = canvasWidth;
            //chartDiv.height = canvasHeight;
			addEl(sdiv, chartDiv);
			var samples = data['samples'];
            var labelLenCutoff = Math.floor((divWidth - 150) / 30);
            var initDrawNum = 10;
            var origSamples = [];
            var initSamples = [];
            var nextSamples = [];
            for (var i = 0; i < samples.length; i++) {
                var sample = samples[i];
                origSamples.push(sample);
                if (sample.length > labelLenCutoff) {
                    sample = sample.substring(0, 4) + '..' + sample.substring(sample.length - 4, sample.length);
                }
                if (i < initDrawNum) {
                    initSamples.push(sample);
                } else {
                    nextSamples.push(sample);
                }
                samples[i] = sample;
            }
			var sos = data['sos'];
			var socountdata = data['socountdata'];
			var datasets = [];
            var initDatasets = [];
            var nextDatasets = [];
			for (var i = 0; i < sos.length; i++) {
				var so = sos[i];
				row = {};
				var label = so
				var backgroundColor = colorPalette[so];
                var counts = socountdata[so];
                var initDatasetCounts = [];
                var nextDatasetCounts = [];
                for (var j = 0; j < counts.length; j++) {
                    if (j < initDrawNum) {
                        initDatasetCounts.push(counts[j]);
                    } else {
                        nextDatasetCounts.push(counts[j]);
                    }
                }
				row['data'] = socountdata[so];
                initDatasets.push({'label': label, 'backgroundColor': backgroundColor, 'data': initDatasetCounts});
                nextDatasets.push({'label': label, 'backgroundColor': backgroundColor, 'data': nextDatasetCounts});
			}
			var chart = new Chart(chartDiv, {
				type: 'bar',
				data: {
					labels: initSamples,
					datasets: initDatasets
				},
				options: {
					title: {
						display: true,
					},
					tooltips: {
						mode: 'index',
						intersect: false,
                        callbacks: {
                            title: function (tooltipItem) {
                                var title = origSamples[tooltipItem[0].index];
                                return title;
                            }
                        },
					},
                    legend: {
                        position: 'left',
                    },
					responsive: true,
                    responsiveAnimationDuration: 500,
                    maintainAspectRatio: false,
					scales: {
						xAxes: [{
                            scaleLabel: {
                                display: true,
                                labelString: 'Samples',
                            },
							stacked: true,
                            ticks: {
                                maxRotation: 90,
                                minRotation: 90,
                            }
						}],
						yAxes: [{
                            scaleLabel: {
                                display: true,
                                labelString: 'Fraction of Sequence Ontology',
                            },
							stacked: true
						}]
					},
				},
			});
            setTimeout(function () {
                for (var j = 0; j < nextSamples.length; j++) {
                    chart.data.labels.push(nextSamples[j]);
                }
                for (var i = 0; i < chart.data.datasets.length; i++) {
                    for (var j = 0; j < nextDatasets[i].data.length; j++) {
                        chart.data.datasets[i].data.push(nextDatasets[i].data[j]);
                    }
                }
                var newWidth = Math.max(divWidth, $(sdiv).width() + nextSamples.length * 20);
                $(sdiv).width(newWidth);
            }, 100);
		}
	}
};

File: webviewerwidgets/wgsosamplesummary/wgsosamplesummary.py
import aiosqlite
import os
import json

async def get_data (queries):
    '''
    so_dic = {
        None: 'Intergenic',
        '':'Intergenic',
        'MIS':'Missense',
        'SYN':'Synonymous',
        'FSI':'Frameshift insertion',
        'FI1':'Frameshift insertion',
        'FI2':'Frameshift insertion',
        'FSD':'Frameshift deletion',
        'FD1':'Frameshift deletion',
        'FD2':'Frameshift deletion',
        'INI':'Inframe insertion',
        'IIV':'Inframe insertion',
        'IND':'Inframe deletion',
        'IDV':'Inframe deletion',
        'CSS':'Complex substitution',
        'STG':'Stopgain',
        'STL':'Stoploss',
        'SPL':'Splice site',
        '2KU':'2k upstream',
        '2KD':'2k downstream',
        'UT3':'3\' UTR',
        'UT5':'5\' UTR',
        'INT':'Intron',
        'UNK':'Unknown',
        'missense':'Missense',
        'synonymous':'Synonymous',
        'frameshift insertion':'Frameshift insertion',
        'frameshift insertion by 1':'Frameshift insertion',
        'frameshift insertion by 2':'Frameshift insertion',
        'frameshift deletion':'Frameshift deletion',
        'frameshift deletion by 1':'Frameshift deletion',
        'frameshift deletion by 2':'Frameshift deletion',
        'inframe insertion':'Inframe insertion',
        'inframe deletion':'Inframe deletion',
        'complex substitution':'Complex substitution',
        'stop gained':'Stopgain',
        'stop lost':'Stoploss',
        'splice site':'Splice site',
        '2kb upstream':'2k upstream',
        '2kb downstream':'2k downstream',
        '3-prime utr':'3\' UTR',
        '5-prime utr':'5\' UTR',
        'intron':'Intron',
        'unknown':'Unknown'
    }
    '''
    
    dbpath = queries['dbpath']
    conn = await aiosqlite.connect(dbpath)
    cursor = await conn.cursor()

    q = 'select distinct base__sample_id from sample where base__sample_id is not null'
    await cursor.execute(q)
    samples = [v[0] for v in await cursor.fetchall() if v[0]]
    if len(samples) == 1:
        response = {'data': None}
        return response
    samples.sort()
    
    q = 'select subdict from variant_reportsub where module="base"'
    await cursor.execute(q)
    r = await cursor.fetchone()
    so_dic = json.loads(r[0])['so']
    so_dic[None] = 'Intergenic'
    so_dic[''] = 'Intergenic'
    q = 'select distinct variant.base__so from variant, variant_filtered where variant.base__uid=variant_filtered.base__uid'
    await cursor.execute(q)
    sos = [so_dic[v[0]] for v in await cursor.fetchall()]
    sos.sort()
    
    sosample = {}
    for so in sos:
        sosample[so] = []
        for sample in samples:
            sosample[so].append(0)
    
    for i in range(len(samples)):
        sample = samples[i]
        q = 'select variant.base__so, count(*) from variant, variant_filtered, sample where variant.base__uid=variant_filtered.base__uid and sample.base__uid=variant.base__uid and sample.base__sample_id="' + sample + '" group by variant.base__so order by variant.base__so'
        await cursor.execute(q)
        for row in await cursor.fetchall():
            (so, count) = row
            so = so_dic[so]
            sosample[so][i] = count

    data = {}
    for so in sos:
        row = []
        for i in range(len(samples)):
            row.append(sosample[so][i])
        data[so] = row
    
    response = {'data': {'samples': samples, 'sos': sos, 'socountdata': data}}

    return response
